### ISSUE TITLE
Compat.stderr != STDERR

### ISSUE THREAD

**Issue by cstjean** [JuliaLang/IJulia.jl#654]
I've been seeing some weird buffering in IJulia recently (notably with ProgressMeter.jl, which uses `Compat.stderr`), and I suspect it's because `Compat.stderr == STDERR` is false in IJulia on 0.6 (but true at the REPL). Likewise for STDOUT/STDIN.

**Comment by stevengj** [comment]
It seems there are a lot of problems with redirection interacting with code that copies a reference to `stderr`.  See also e.g. JuliaLang/julia#26798.

I wonder if a more general solution would be to make `stdout` and `stderr` in Julia a mutable container along the lines of:
```
mutable type IOContainer
    io::IO
end
const stdout = IOContainer(...real stdout object...)
```
with `redirect_stdout` mutating `stdout.io = ...new IO...`.   That way existing references to `stdout` would be updated.

**Comment by stevengj** [comment]
But in the meantime IJulia probably needs to rework how it redirects I/O.

**Comment by stevengj** [comment]
@cstjean, have you already checked whether this was fixed on IJulia master by #641?

**Comment by cstjean** [comment]
Yes, that's the version I'm on.

**Comment by stevengj** [comment]
Looks like the problem is that #641 hard-coded `redirect_stdout` to directly change `STDOUT`.

**Comment by cstjean** [comment]
Might be unrelated, but this looks wrong:

![screen shot 2018-04-19 at 21 35 30](https://user-images.githubusercontent.com/8622776/39026207-c0773334-4419-11e8-912f-437a21094c91.png)

I don't know how `Compat.stdout` manages to track STDOUT at the REPL inside of `redirect_stdout`, but whatever it's doing is broken in IJulia.

**Comment by cstjean** [comment]
AFAICT, this was fixed by #671 / not relevant on 1.0
